objects/src/accounts/vault.rs

```diff
@@ -0,0 +1,68 @@
+use super::{AccountError, AccountId, Asset, Digest};
+
+// ACCOUNT VAULT
+// ================================================================================================
+
+/// An asset container for an account.
+///
+/// An account vault can contain an unlimited number of assets. The assets are stored in a Sparse
+/// Merkle tree as follows:
+/// - For fungible assets, the index of a node is defined by the issuing faucet ID, and the value
+///   of the node is the asset itself. Thus, for any fungible asset there will be only one node
+///   in the tree.
+/// - For non-fungible assets, the index is defined by the asset itself, and the asset is also
+///   the value of the node.
+///
+/// An account vault can be reduced to a single hash which is the root of the Sparse Merkle tree.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct AccountVault {
+    // TODO: add backing sparse Merkle tree
+    assets: Vec<Asset>,
+}
+
+impl AccountVault {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new account vault initialized with the provided assets.
+    ///
+    /// TODO: return error if there are duplicates in the provided asset list.
+    pub fn new(assets: &[Asset]) -> Self {
+        Self {
+            // TODO: put assets into a Sparse Merkle trees
+            assets: assets.to_vec(),
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a commitment to this vault.
+    pub fn root(&self) -> Digest {
```

Comment:
the root read from the mekle backend might fail as it might have to be fetched from the storage. if we want to enforce it to always exist in-memory (which is a reasonable requirement), we might want to return a reference to a `Digest` here (i.e. `pub const fn root(&self) -> &Digest`

Reply:
I'm not sure I fully understand this. If retrieval of a root fails, wouldn't this panic? Also, I'd rather keep this infallible for now as I'm not sure how it'll fit the bigger system.